### ISSUE TITLE
Multiple element is sharing the variable if no initial variable assign to tags property.

### ISSUE THREAD

**Issue by cheonhyangzhang** [cheonhyangzhang/paper-tags-input#13]
Per @apalaniuk 
it's well worth noting here that your default value initialization of the tags property doesn't seem right. You're setting the value to be an empty array, which will cause a single array to be used across all instances of the component, so you can't use multiple components on the same page with independent tags. The default value for arrays and objects (if set) should be a function that returns a new array instead, which is invoked once per instance - or you can do as you're doing in _addTag and lazily initialize if needed. See https://www.polymer-project.org/1.0/docs/devguide/properties#configure-values

Also, if (typeof(this.tags) == 'undefined' || this.tags == null) is redundant - this.tags == null will catch both null and undefined. But, shouldn't it be this.tags instanceof Array?

**Comment by apalaniuk** [comment]
@cheonhyangzhang @barkerb1
I should have said, I think the `typeof` check is more accurate for a defensive check, but I'm not sure if there should be a check at all. If `tags` is not an array, AFAICT it's because the consumer has accidentally bound something other than an array to that property, so exploding may be the best thing to do there. Just a thought.

Or perhaps @barkerb1 has found a case where `tags` can be `null` not as a result of client code?

**Comment by cheonhyangzhang** [comment]
@apalaniuk  
Can you think of any of the existing element Polymer ( https://elements.polymer-project.org/browse?package=paper-elements ) that has the similar situation we have here? So that maybe we can see what Google is doing in this case.

**Comment by apalaniuk** [comment]
@cheonhyangzhang 

I don't think Polymer elements are designed the same way. For instance, if you look at [`iron-selectable`](https://github.com/PolymerElements/iron-selector/blob/master/iron-selectable.html#L349) which underpins `iron-menu-behavior`, `paper-listbox`, `paper-menu`, etc. which I think is a reasonably close example, they will use [mutation observers](https://www.polymer-project.org/1.0/docs/devguide/local-dom#observe-nodes) to listen to changes to the light DOM, query children based on some given selector, and update internal state based on that. (Other than the documentation, note that there's a [`polymer-patterns`](https://github.com/PolymerLabs/polymer-patterns) repo that has examples of idiomatic Polymer, [including mutation observers](https://github.com/PolymerLabs/polymer-patterns/blob/master/snippets/observing-changes/observing-changes-to-light-dom-children.html)).

But more generally, if your contract states that a public property is an array, and you initialize with a valid default value, I think the best thing you can do is fail early and loudly if the consumer violates that, rather than carry on silently with unknown behavior. If the Polymer team did the same checks, they would have an observer on every public property with `typeof` or `instanceof` checks and exceptions and whatnot, and they don't.

**Comment by barkerb1** [comment]
So in my use case I can implement the tag simply with an empty array without an issue:
`<paper-tags-input tags="[]"></paper-tags-input>`

My personal design with my own elements is to always include conditions to check for any mandatory properties before execution of any logic.  Since Polymer's error handling leaves much to be desired for debugging, I'll throw a vary specific error for each case that fails.  
Such as: `throw 'Error: tags property not initialized as an Array, typeof tags= ' + typeof tags`

**Comment by cheonhyangzhang** [comment]
@barkerb1  @apalaniuk 
Actually I found that current implementation of paper-tags-input is already throwing out error in the console

``` javascript
VM61 polymer.html:3701 [dom-repeat::dom-repeat]: expected array for `items`, found hello
```

And it's already easy enough to figure out the reason is items property is expecting array. So as for the typeof tags problem I think we should already be good.

**Comment by cheonhyangzhang** [comment]
This is handled by #19

**Comment by cheonhyangzhang** [comment]
@barkerb1 @apalaniuk 

The changes are implemented in v1.2.4
https://github.com/cheonhyangzhang/paper-tags-input/releases/tag/v1.2.4
